packages/ai/src/registry.ts

```diff
@@ -12,15 +12,112 @@ import type {
   ModelCapability,
   ModelHandler,
   ModelInfo,
+  ModelMetrics,
   ModelOverride,
+  ModelPricing,
   ModelSelectionWeights,
   ModelSource,
   Provider,
   Providers,
+  RangeConstraint,
   ScoredModel,
   SelectedModel
 } from './types';
 
+// ============================================================================
+// Generic Constraint Helpers
+// ============================================================================
+
+/**
+ * Extracts a numeric value from a complex object for comparison.
+ * Supports nested paths and averaging of multiple values.
+ */
+function extractNumericValue(
+  value: number | ModelPricing | ModelMetrics | undefined,
+  path?: string
+): number | undefined {
+  if (value === undefined) return undefined;
+  if (typeof value === 'number') return value;
+  
+  // For ModelPricing, average input/output text costs
+  if ('text' in value) {
+    const pricing = value as ModelPricing;
+    const input = pricing.text?.input;
+    const output = pricing.text?.output;
+    // If neither input nor output pricing is defined, return undefined
+    if (input === undefined && output === undefined) return undefined;
+    // Calculate average of available values
+    return ((input ?? 0) + (output ?? 0)) / 2;
+  }
+  
+  // For ModelMetrics, use tokensPerSecond as primary metric
+  if ('tokensPerSecond' in value) {
+    const metrics = value as ModelMetrics;
+    return metrics.tokensPerSecond;
+  }
+  
+  return undefined;
+}
+
+/**
+ * Checks if a model value satisfies a range constraint.
+ * Returns true if the value is acceptable, false otherwise.
+ */
+function satisfiesConstraint<T>(
```

Comment:
Should be:
- modelValue: T | undefined
- min: T | undefined
- max: T | undefined

And if T is an object it should recursively call satisfiesConstraint to cover all properties of T. extractNumericValue is a bad approach